codex/erasure/erasure.nim

```diff
@@ -269,6 +287,52 @@ proc init*(
     strategy: strategy,
   )
 
+proc leopardEncodeTask(tp: Taskpool, task: ptr EncodeTask) =
+  # Task suitable for running in taskpools - look, no GC!
+  if (let res = task[].encoder.encode(task[].blocks[], task[].parity[]); res.isErr):
+    warn "Error from leopard encoder backend!", error = $res.error
+    task[].success.store(false)
+  else:
+    task[].success.store(true)
+
+  discard task[].signal.fireSync()
+
+proc encodeAsync(
+    self: Erasure,
+    blockSize, ecK, ecM: int,
+    data: ref seq[seq[byte]],
+    parity: ptr seq[seq[byte]],
+): Future[?!void] {.async: (raises: []).} =
+  ## Create an encoder backend instance
+  let encoder = self.encoderProvider(blockSize, ecK, ecM)
+  without threadPtr =? ThreadSignalPtr.new():
+    return failure("Unable to create thread signal")
+
+  ## Create an ecode task with block data 
+  var task =
+    EncodeTask(encoder: encoder, blocks: data, parity: parity, signal: threadPtr)
+
+  let t = addr task
+  doAssert self.taskPool.numThreads > 1,
+    "Must have at least one separate thread or signal will never be fired"
+  self.taskPool.spawn leopardEncodeTask(self.taskPool, t)
+
+  try:
+    await threadPtr.wait()
+  except AsyncError as exc:
+    warn "Async thread error", err = exc.msg
+    return failure(exc.msg)
+  except CancelledError:
+    return failure("Thread cancelled")
```

Comment:
Here, do we want to return a failure? I thought about this a bit further... this proc executes in the master thread, correct? If so, maybe we should propagate `CancelledErrors`. So the raises annotation would be:

```nim
{.async: (raises: [CacncelledError]).}
```
and the except block would re-raise:
```nim
except CancelledError as e:
    raise e
```

This brings up another question -- should the `TaskPool` be gracefully terminated in `stop`? Or is there no point to wait for worker threads to join master? IOW, would stopping the master thread kill the worker at the OS-level and possibly produce an out-of-band error?

Reply:
Hmm, couldn't that create a deadlock situation where the task is cancelled with `cancelAndWait` and but cancellation has wait on the thread to finish its operation before cancelling? Is there a way to forcefully kill the task and discard any errors or results?